### PR TITLE
les: add checkpoint challenge for light client

### PR DESCRIPTION
**Pull request by rjl493456442**
Replace #20125

### CODE REVIEW

**Comment by rjl493456442** [comment]
This PR is blocked by #20692

**Comment by rjl493456442** [comment]
I should rebase it and @zsfelfoldi Please take a look.

**Comment by holiman** [comment]
Discussion: instead of matching wildcard responses, use the request-id / response id for matching.

File: les/client_handler.go [hunk 1]
```diff
@@ -205,6 +232,27 @@ func (h *clientHandler) handleMsg(p *serverPeer) error {
...
+
+		// If no headers were received, but we're expecting a checkpoint header,
+		// drop the unsynced server.
+		if len(resp.Headers) == 0 && p.syncDrop != nil {
+			p.syncDrop.Stop()
+			p.syncDrop = nil
+			return errResp(ErrUselessPeer, "msg %v: %v", msg, err)
+		}
+		// If we are still waiting the checkpoint response.
+		if len(resp.Headers) == 1 && p.syncDrop != nil {
```
Comment (rjl493456442) [inline les/client_handler.go:244]: Simplify it by checking the reqid

**Comment by rjl493456442** [comment]
@zsfelfoldi Think about this PR for a while, now I am inclined to close this PR.

The main purpose of the checkpoint is we want to ensure the remote peer is in the same chain as us before we run the sync protocol. It's super useful for ETH protocol because we sync from the local head anyway and it will take a very long time to finish the sync(usually).

However, for LES, the checkpoint challenge seems unnecessary. Light client syncs from the checkpoint anyway and the checkpoint is verified in two ways:
- It's the hardcoded checkpoint, trust it anyway, no verification
- It's the signed checkpoint by Geth devs, verify the signature
And after the verification, we will firstly sync the checkpoint and then the remaining headers.
So it means before the real sync, we already ensure the remote peer is on the same side as us.

It might be useful in the future if we implement the flexible light sync that can start in any specified CHT. In this case, users want the legacy headers but still want to sync the correct chain. So the checkpoint challenge can help.
But really not sure whether it's worthwhile to add this complexity.

**Pull request closed without merging**